server/api_archive.go

```diff
@@ -0,0 +1,170 @@
+	if r.Method != http.MethodGet {
+		http.Error(w, "Only GET method is allowed", http.StatusMethodNotAllowed)
+		return
+	}
+
+	if r.ContentLength != 0 {
+		http.Error(w, "Request should not have a body", http.StatusBadRequest)
+		return
+	}
+
+	if !gcas.ApiArchiveRateLimiter.Allow() {
+		http.Error(w, fmt.Sprintf("Too many requests, this server allows %v every %v", apiArchiveLimit, apiArchiveRate), http.StatusTooManyRequests)
+		return
+	}
+
+	buffer := new(bytes.Buffer)
+	znw := zip.NewWriter(buffer)
+
+	// File ordering matters here. We are not locking the files, and relying on the file writes being
+	// append only, and happening in a single write. We must also archive files later which depend
+	// on earlier files.
+
+	// Add the pseudo file server.pubkey
+	if err := gcas.addPubKeyFile("server.pubkey", znw); err != nil {
+		http.Error(w, fmt.Sprintf("Error zipping server.pubkey: %v", err), http.StatusInternalServerError)
+		return
+	}
+
+	for _, name := range PublicFiles() {
+		err := gcas.addFile(name, znw)
+		if err != nil {
+			http.Error(w, fmt.Sprintf("Error zipping file %v: %v", name, err), http.StatusInternalServerError)
+			return
+		}
+	}
+
+	// Add a README
+	if err := gcas.addReadmeFile(znw); err != nil {
+		http.Error(w, fmt.Sprintf("Error adding README: %v", err), http.StatusInternalServerError)
+		return
+	}
+
+	if err := znw.Close(); err != nil {
+		if _, err := w.Write(buffer.Bytes()); err != nil {
+			http.Error(w, "Failed to close zip writer", http.StatusInternalServerError)
+			return
+		}
+	}
+
+	w.Header().Set("Content-Type", "application/zip")
+
+	if _, err := w.Write(buffer.Bytes()); err != nil {
+		http.Error(w, "Failed to write archive response", http.StatusInternalServerError)
+	}
+}
+
+func (gcas *GCAServer) addFile(name string, zipw *zip.Writer) error {
+	path := filepath.Join(gcas.baseDir, name)
+
+	file, err := os.Open(path)
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+
+	fileInfo, err := file.Stat()
+	if err != nil {
+		return err
+	}
+
+	header, err := zip.FileInfoHeader(fileInfo)
+	if err != nil {
+		return err
+	}
+
+	header.Method = zip.Deflate
+
+	writer, err := zipw.CreateHeader(header)
+	if err != nil {
+		return err
+	}
+
+	// File contains only public data so copy it all.
+	if _, err := io.Copy(writer, file); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (gcas *GCAServer) addPubKeyFile(name string, znw *zip.Writer) error {
+	path := filepath.Join(gcas.baseDir, "server.keys")
+
+	file, err := os.Open(path)
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+
+	fileInfo, err := file.Stat()
+	if err != nil {
+		return err
+	}
+
+	writer, err := znw.CreateHeader(&zip.FileHeader{
+		Name:     "server.pubkey",
+		Method:   zip.Deflate,
+		Modified: fileInfo.ModTime(),
+	})
+
+	if err != nil {
+		return err
+	}
+
+	// Copy only the public key part of the file
+	if _, err := io.CopyN(writer, file, 32); err != nil {
```

Comment:
Can you double and triple check that the first 32 bytes are the public bytes?

Reply:
I think I can make it more explicit by copying the public key part of the structure